include/util/cell_storage.hpp

```diff
@@ -0,0 +1,295 @@
+        {
+            return std::find(
+                       destination_boundary, destination_boundary + num_destination_nodes, node) -
+                   destination_boundary;
+        }
+
+      public:
+        std::pair<RowIterator, RowIterator> GetOutWeight(NodeID node) const
+        {
+            auto row = GetRow(node);
+            auto begin = weights + num_destination_nodes * row;
+            auto end = begin + num_destination_nodes;
+            return std::make_pair(begin, end);
+        }
+
+        std::pair<ColumnIterator, ColumnIterator> GetInWeight(NodeID node) const
+        {
+            auto column = GetColumn(node);
+            auto begin = ColumnIterator{weights + column, num_destination_nodes};
+            auto end =
+                ColumnIterator{weights + column + num_source_nodes * num_destination_nodes,
+                               num_destination_nodes};
+            return std::make_pair(begin, end);
+        }
+
+        auto GetSourceNodes() const
+        {
+            return std::make_pair(source_boundary, source_boundary + num_source_nodes);
+        }
+
+        auto GetDestinationNodes() const
+        {
+            return std::make_pair(destination_boundary,
+                                  destination_boundary + num_destination_nodes);
+        }
+
+        CellImpl(const CellData &data,
+                 WeightPtrT const all_weight,
+                 const NodeID *const all_sources,
+                 const NodeID *const all_destinations)
+            : num_source_nodes{data.num_source_nodes},
+              num_destination_nodes{data.num_destination_nodes},
+              weights{all_weight + data.weight_offset},
+              source_boundary{all_sources + data.source_boundary_offset},
+              destination_boundary{all_destinations + data.destination_boundary_offset}
+        {
+        }
+    };
+
+    inline std::size_t LevelIDToIndex(LevelID level) const { return level-1; }
+
+  public:
+    using Cell = CellImpl<EdgeWeight *>;
+    using ConstCell = CellImpl<const EdgeWeight *>;
+
+    template <typename GraphT>
+    CellStorage(const MultiLevelPartition &partition, const GraphT &base_graph)
+    {
+        // pre-allocate storge for CellData so we can have random access to it by cell id
+        unsigned number_of_cells = 0;
+        for (LevelID level = 1u; level < partition.GetNumberOfLevels(); ++level)
+        {
+            level_to_cell_offset.push_back(number_of_cells);
+            number_of_cells += partition.GetNumberOfCells(level);
+        }
+        level_to_cell_offset.push_back(number_of_cells);
+        cells.resize(number_of_cells);
+
+        for (LevelID level = 1u; level < partition.GetNumberOfLevels(); ++level)
+        {
+            auto level_offset = level_to_cell_offset[LevelIDToIndex(level)];
+
+            std::vector<std::pair<CellID, NodeID>> level_source_boundary;
+            std::vector<std::pair<CellID, NodeID>> level_destination_boundary;
```

Comment:
Hm this creates vectors and destroys them after each iteration again. Should be fine with only a couple of levels, but still you could hoist them outside and `.clear()` between each iteration instead.